common/Storyboarding/OsbAnimationWriter.cs

```diff
@@ -0,0 +1,128 @@
+            {
+                //this shouldn't loop again so we need a sprite instead
+                var sprite = new OsbSprite()
+                {
+                    InitialPosition = OsbAnimation.InitialPosition,
+                    Origin = OsbAnimation.Origin,
+                    TexturePath = GetLastFramePath(),
+                };
+
+                foreach (var command in segment)
+                    sprite.AddCommand(command);
+
+                return sprite;
+            }
+            else
+            {
+                var animation = new OsbAnimation()
+                {
+                    TexturePath = OsbAnimation.TexturePath,
+                    InitialPosition = OsbAnimation.InitialPosition,
+                    Origin = OsbAnimation.Origin,
+                    FrameCount = OsbAnimation.FrameCount,
+                    FrameDelay = OsbAnimation.FrameDelay,
+                    LoopType = OsbAnimation.LoopType,
+                };
+
+                foreach (var command in segment)
+                    animation.AddCommand(command);
+
+                return animation;
+            }
+        }
+
+        private string GetLastFramePath()
+        {
+            var dir = Path.GetDirectoryName(OsbAnimation.TexturePath);
+            var file = string.Concat(Path.GetFileNameWithoutExtension(OsbAnimation.TexturePath), OsbAnimation.FrameCount - 1, Path.GetExtension(OsbAnimation.TexturePath));
+            return Path.Combine(dir, file);
+        }
+
+        protected override void WriteHeader(OsbSprite sprite)
+        {
+            if (sprite is OsbAnimation animation)
+            {
+                double frameDelay = animation.FrameDelay;
+                TextWriter.WriteLine($"Animation,{OsbLayer},{animation.Origin},\"{OsbSprite.TexturePath.Trim()}\",{animation.InitialPosition.X.ToString(ExportSettings.NumberFormat)},{animation.InitialPosition.Y.ToString(ExportSettings.NumberFormat)},{animation.FrameCount},{frameDelay.ToString(ExportSettings.NumberFormat)},{animation.LoopType}");
+            }
+            else
+            {
+                base.WriteHeader(sprite);
+            }           
+        }
+
+        protected override HashSet<int> GetFragmentationTimes()
+        {
+            HashSet<int> fragmentationTimes = base.GetFragmentationTimes();
+
+            int tMax = fragmentationTimes.Max();
+            HashSet<int> nonFragmentableTimes = new HashSet<int>();
+
+            for (double d = OsbAnimation.StartTime; d < OsbAnimation.AnimationEndtime(); d += OsbAnimation.LoopDuration())
+            {
+                var range = Enumerable.Range((int)d + 1, (int)(OsbAnimation.LoopDuration() - 1));
+                nonFragmentableTimes.UnionWith(range);
+            }
+
+            fragmentationTimes.RemoveWhere(t => nonFragmentableTimes.Contains(t) && t < tMax);
+
+            return fragmentationTimes;
+        }
+    }
+
+    static class OsbAnimationExtensions

```

Comment:
No need for extensions, put these in OsbAnimation  as one line properties (=>)

Reply:
makes sense, done